### ISSUE TITLE
Opening a file.

### ISSUE THREAD

**Issue by dam5s** [elmish/Elmish.WPF#96]
I'm trying to hook an Microsoft.Win32.OpenFileDialog into the application.

If I had access to the dispatch I probably would build a solution similar to this:

```
   mainWindow.Dispatcher.Invoke(fun () ->
        let dlg = Microsoft.Win32.OpenFileDialog ()
        dlg.DefaultExt <- ".epub"
        dlg.Filter <- "EPUB Files (*.epub)"

        let result = dlg.ShowDialog ()

        let newBook =
            if result.HasValue && result.Value
            then Book.fromPath dlg.FileName
            else Error "No book selected"

        dispatch (FileSelectResult newBook)
    )
```

But I don't see a way to achieve this.

I saw in other dialog examples that the dialog could share the same `DataContext` so that we can apply bindings to it. But it does not look that it would be possible with the `OpenFileDialog`.

Is there a recommended way to deal with this?

**Comment by et1975** [comment]
If you need a dispatch then you probably want to create a custom `Cmd`, something like:
```fsharp
module Dialog =
    let openFile mainWindow onOK onCancel =
        [fun dispatch -> (* your Invoke *)]
```
and used like this
```fsharp
let update msg model =
   model, Dialog.openFile FileSelectResult FileSelectCancelled
```

**Comment by cmeeren** [comment]
Yes, if you need `dispatch` then making a custom `Cmd` seems to be the right option.

Note that (AFAIK) you can also use `Application.Current.Dispatcher` instead of passing a window and use that.

Also note that I've just released 3.2.0 with proper support for new windows (modal and non-modal), though that's for `Window`s, not `OpenFileDialog`.

**Comment by dam5s** [comment]
Great, thanks for the quick feedback to both of you!
